conversant/prompt_chatbot.py

```diff
@@ -244,56 +244,76 @@ def generate_prompt_update_examples(self, query: str) -> str:
-        is_final_chunk = False
-
-        stop_seq = self.should_stop(response)
-        if stop_seq != None or response == "":
-            is_final_chunk = True
-            if stop_seq != None:
-                response = response[: -len(stop_seq)]
-
-        if is_from_scratch:
-            response = response.lstrip()
+        if add_new_elem:
             self.chat_history.append(self.prompt.create_interaction(query, response))
             self.prompt_size_history.append(
                 self.co.tokenize(
                     self.prompt.create_interaction_string(query, response)
                 ).length
             )
             self.prompt_history.append(current_prompt)
-
         elif response != "":
             self.chat_history[-1]["bot"] += response
             self.prompt_history[-1] += response
             self.prompt_size_history[-1] += self.co.tokenize(response).length
 
-        return self.chat_history[-1]["bot"], is_final_chunk
+    def partial_reply(self, query: str) -> str:
+        """Generates (partial) reply to a query given a chat history.
+
+        Args:
+            query (str): A query passed to the prompt chatbot.
+            is_from_scratch (bool): Tells the chatbot if the reply
+            should be generated from scratch or it is the continuation
+            of the previous chunk
+
+        Returns:
+            str: Dictionary of query and generated LLM response
+        """
+        current_prompt = self.generate_prompt_update_examples(query)
+        response_so_far = ""
+        for i in range(self.partial_reply_max_reruns()):
+            generated_object = self.co.generate(
+                model=self.client_config["model"],
+                prompt=current_prompt,
+                max_tokens=TOKENS_PER_REQUEST,
+                temperature=self.client_config["temperature"],
+                frequency_penalty=self.client_config["frequency_penalty"],
+                presence_penalty=self.client_config["presence_penalty"],
+                stop_sequences=self.client_config["stop_sequences"],
+            )
+            response = generated_object.generations[0].text
+            yield response
+            # If first chunk, remove leading whitespace
+            if i == 0:
+                response = response.lstrip()
+                self.append_to_chat_history(query, response, current_prompt, True)
+
+            stop_seq = self.should_stop(response)
+            if stop_seq != "" or response == "":
+                if stop_seq != "":
+                    response = response[: -len(stop_seq)]
+                    self.append_to_chat_history(query, response, current_prompt, False)
+                    response_so_far += response
+                break
+            current_prompt += response
+            response_so_far += response
+            if i != 0:
+                self.append_to_chat_history(query, response, current_prompt, False)
+            yield response_so_far
```

Comment:
I think I'm a little confused about this chunk of code, and it's a bit hard for me to follow - in particular, why do we have two `yield` statements in one iteration of the for loop, one for `response`, and another for `response_so_far`? 

Perhaps we can comment these few lines to make what is happening more explicit

Reply:
similar question from me -- I see in the streamlit when we make the generator we immediately exhaust the first yield statement by calling next(), is that right?

It seems like this function yields both the isolated reply chunks and the reply-so-far in alternation; IMO that could get dicey when calling this code. Is it possible to write this function in a way that it only does one of these behaviours (and we should also note that behaviour in the docstring for the function)?

Reply:
OH that was a typo, I removed it!! Thank you for catching that! Also, I cleaned it up a little bit, let me know what you think!